### DISCUSSION TITLE
I want to take orders

### DISCUSSION THREAD

**Discussion by MustafaAlperen** [etsy/open-api#983] [Q&A] [answered]
Hello,
I am a new developer for ETSY.
In my project that I developed with ASP.NET MVC, I receive tokens with OAuth. AR then I want to pull orders with token, shop id and app keystring. I wrote my code as below, but I can't get any results. Where am I doing wrong? Can you help me? 

I am getting 403 error.

I get transaction authorizations with the following scope while logging in.

scope=transactions_w%20transactions_r%20address_r%20address_w%20billing_r%20cart_r%20cart_w%20email_r%20favorites_r%20favorites_w%20feedback_r%20listings_d%20listings_r%20listings_w%20profile_r%20profile_w%20recommend_r%20recommend_w%20shops_r%20shops_w

```
public async Task<List<JToken>> Transactions()
{
    string accessToken = CurrentSession.EtsyToken; //Get from Oauth token
    int shopId = WebConfigHelper.EtsyShopId;
    string apiKey = WebConfigHelper.EtsyKeystring; // ETSY key string

    using (var handler = new HttpClientHandler())
    {
        handler.SslProtocols = SslProtocols.Tls12 | SslProtocols.Tls11 | SslProtocols.Tls;

        using (HttpClient client = new HttpClient(handler))
        {
            client.BaseAddress = new Uri("https://openapi.etsy.com/");
            client.DefaultRequestHeaders.Authorization = new AuthenticationHeaderValue("Bearer", accessToken);
            client.DefaultRequestHeaders.Add("x-api-key", apiKey);

            HttpResponseMessage response = await client.GetAsync($"v3/application/shops/{shopId}/transactions");

            if (response.IsSuccessStatusCode)
            {
                string jsonResponse = await response.Content.ReadAsStringAsync();
                JObject jObject = JObject.Parse(jsonResponse);
                JToken results = jObject["results"];

                return results.ToObject<List<JToken>>();
            }
            else
            {
                Console.WriteLine($"Error: {response.StatusCode}");
                throw new Exception($"Error: {response.StatusCode}");
            }
        }
    }
}
```

**Comment by Gareth064** [comment]
Hey. Couple of things here.

A transaction is not the order, it's the line item of an order.

To get Orders, then you need to hit this URL


https://openapi.etsy.com/v3/application/shops/{shop_id}/receipts<https://openapi.etsy.com/v3/application/shops/%7Bshop_id%7D/receipts>


Secondly, 403 (Forbidden) means you likely haven't got the right scopes for your Auth token, you need to make sure you generated your Auth token with the scope `transactions_r`

**Reply by Gareth064** [reply]
I am in the middle of building a .Net Library Wrapper around the Etsy API. It does all the gets of listing and receipt endpoints so far, as well as handle Auth. I might look at releasing it on github as a pre-release thing or something.

**Reply by MustafaAlperen** [reply, marked as answer]
Thank you for your help. Actually, I was entering the shop id incorrectly. :) Using .NET, I am now able to login and receive tokens and receive Receipts. I also extracted it from JSON and converted it to object based. I would like to support your library. If you share, I'll wait at the bottom of this post.

**Reply by Gareth064** [reply]
I'm away on holiday at the moment but when I get back I will look to see how I can get my library into nuget.

**Reply by Archieversam** [reply]
Are you looking for a way to track and convert majority of your traffic into potential buyers in an easy way? Kindly contact us now for more details or use the link below. THANKS.
https://cutt.ly/Sw0vSP70

**Comment by muratali016** [comment]
Hey @Gareth064 how can I change the discount rate for my items via api v3? Is this possible?

**Reply by MustafaAlperen** [reply]
MErhaba,

Aşağıdaki kodlarım işinizi görecektir. BElki http protokol hatası verir bu durumda 

System.Security.Authentication.SslProtocols.Tls12 | System.Security.Authentication.SslProtocols.Tls11 | System.Security.Authentication.SslProtocols.Tls

kullanın.

using System;
using System.Net.Http;
using System.Net.Http.Headers;
using System.Text;
using System.Threading.Tasks;

namespace EtsyApiExample
{
    class Program
    {
        private static readonly HttpClient httpClient = new HttpClient();

        static async Task Main(string[] args)
        {
            string apiKey = "";
            string apiSecret = "";
            string accessToken = "";
            string accessSecret = "";
            string shopId = "";
            string saleId = "";

            httpClient.DefaultRequestHeaders.Authorization = new AuthenticationHeaderValue("Bearer", accessToken);

            await UpdateSaleDiscountPercentageAsync(apiKey, apiSecret, accessToken, accessSecret, shopId, saleId, 30);
        }

        private static async Task UpdateSaleDiscountPercentageAsync(string apiKey, string apiSecret, string accessToken, string accessSecret, string shopId, string saleId, int discountPercentage)
        {
            string requestUri = $"https://api.etsy.com/v3/application/shops/{shopId}/sales/{saleId}";
            string content = $"{{\"discount_pct\": {discountPercentage}}}";

            var httpContent = new StringContent(content, Encoding.UTF8, "application/json");

            var response = await httpClient.PutAsync(requestUri, httpContent);

            if (response.IsSuccessStatusCode)
            {
                Console.WriteLine("Sale discount percentage updated successfully.");
            }
            else
            {
                Console.WriteLine($"Error: {response.StatusCode} - {response.ReasonPhrase}");
            }
        }
    }
}

**Reply by muratali016** [reply]
@MustafaAlperen merhaba çok teşekkür ederim yanıt için . Pyhton da kodluyor musunuz biraz yabancı geldi ama, chatgpt den çevirmesini isteyeceğim eğer Python yapmıyorsanız

**Reply by MustafaAlperen** [reply]
Rica ederim.
Bende internetten kodları c#'dan python'a çevirdim.

Bunu kullanmanız için requests kütüphanesi lazım.

```
def update_sale_discount_percentage(api_key, api_secret, access_token, access_secret, shop_id, sale_id, discount_percentage):
    url = f"https://api.etsy.com/v3/application/shops/{shop_id}/sales/{sale_id}"
    headers = {
        "Authorization": f"Bearer {access_token}",
        "Content-Type": "application/json"
    }
    data = {
        "discount_pct": discount_percentage
    }
    
    response = requests.put(url, headers=headers, data=json.dumps(data))

    if response.status_code == 200:
        print("Sale discount percentage updated successfully.")
    else:
        print(f"Error: {response.status_code} - {response.text}")

if __name__ == "__main__":
    api_key = ""
    api_secret = ""
    access_token = ""
    access_secret = ""
    shop_id = ""
    sale_id = ""
    
    update_sale_discount_percentage(api_key, api_secret, access_token, access_secret, shop_id, sale_id, 30)
```

**Reply by Gareth064** [reply]
I have searched the API documentation and there is no /sales endpoint.

I believe it was asked previously, but there is no endpoint to allow you to setup discounts or coupons. You have to do it via the Etsy Website.

**Reply by muratali016** [reply]
Thanks @Gareth064

**Reply by kanika-kjt** [reply]
Can someone please help me to authenticate and get access token in python? I am beginner and don't know more about it. 
Please help

**Reply by Archieversam** [reply]
Are you looking for a way to track and convert majority of your traffic into potential buyers in an easy way? Kindly contact us now for more details or use the link below. THANKS.
https://cutt.ly/Sw0vSP70

**Comment by Archieversam** [comment]
Are you looking for a way to track and convert majority of your traffic into potential buyers in an easy way? Kindly contact us now for more details or use the link below. THANKS.
https://cutt.ly/Sw0vSP70
